drivers: gpio: nrf5: Add support for port P1 in nRF52840 SoC

cvinayak opened the pull request:
Added support for GPIO Port P1 in nRF52840 SoC.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
drivers: gpio: nrf5: Add support for port P1 in nRF52840 SoC

Added support for GPIO Port P1 in nRF52840 SoC.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
boards: nRF5x DK: Fix missing pull up for switches

Fixed missing GPIO_PUD_PULL_UP configuration setting for
push button switches in the nRF51_pca10028, nRF52_pca10040
and nRF52840_pca10056 DK boards.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@mike-scott @carlescufi @anangl tested button sample on nRF52840 (pin 11 with "GPIO_0" and "GPIO_1").

cvinayak commented on drivers/gpio/gpio_nrf5.c:
```diff
@@ -82,7 +78,7 @@ struct gpio_nrf5_data {
 
 	/*@todo: move GPIOTE channel management to a separate module */
-	u32_t gpiote_chan_mask;
+	u32_t *gpiote_chan_mask;
```
agree

cvinayak commented on drivers/gpio/gpio_nrf5.c:
```diff
@@ -406,30 +500,87 @@ void nrf_gpiote_clear_port_event(void)
+	}
+}
+
+static int gpio_port_get(struct device *dev)
```
agree

cvinayak replied:
done

cvinayak commented on drivers/gpio/gpio_nrf5.c:
```diff
@@ -406,30 +500,87 @@ void nrf_gpiote_clear_port_event(void)
+
+#endif /* CONFIG_GPIO_NRF5_P1 */
+
+static inline struct device *gpio_dev_get(int port)
```
I did not want to move the entire isr function down here (hence this function is with inline, and a const port param), This function being inline,  only the const returned by DEVICE_GET gets placed in the isr caller, there is no efficiency compromised.

Do you still want me to move the isr down here? I am neutral.

anangl replied:
I don't like moving isr here, too. What about using `DEVICE_DECLARE`?

cvinayak replied:
done

mike-scott commented:
I understand there are some review comments that need to be addressed, however, tested this patch w/ analyzer hooked up to 6 different pins.   Works as expected w/ P1 GPIOs.

cvinayak commented:
@mike-scott Thank you for testing (just not able to get back to this PR due to other priorities), is this in any way blocking you?

mike-scott commented:
@cvinayak: not blocking me.  Take your time with this. :)

cvinayak committed:
drivers: gpio: nrf5: Use popcount and first_lsb_set functions

Use the Zephyr toolchain and architecture defined popcount
and first_lsb_set functions instead of GCC specific
functions.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak reviewed (commented):
@anangl @mike-scott reworked, ready for your approvals.

mike-scott commented on drivers/gpio/Kconfig.nrf5:
```diff
@@ -27,13 +25,27 @@ config GPIO_NRF5_P0_DEV_NAME
+	range 0 2 if SOC_SERIES_NRF51X
 	range 0 5 if SOC_SERIES_NRF52X
-	default 1
+	default 2 if SOC_SERIES_NRF51X
```
Changing the default interrupt priority (1->2 on nRF51 and 1->5 on nRF52) almost seems like a patch that should land on top of the GPIO driver change to support P1.   It might go unnoticed as a part of this larger patch.  What do you think?

cvinayak replied:
I will make it a separate commit.

mike-scott commented on drivers/gpio/gpio_nrf5.c:
```diff
@@ -386,6 +449,18 @@ static const struct gpio_driver_api gpio_nrf5_drv_api_funcs = {
 	.disable_callback = gpio_nrf5_disable_callback,
 };
 
+static u32_t gpiote_chan_mask;
```
I believe this was defined above at line 69.

cvinayak replied:
good catch, will fix it. Would why i got not compiler error/warning?! map file correctly had a single instance allocated.

mike-scott commented on drivers/gpio/gpio_nrf5.c:
```diff
@@ -422,14 +500,30 @@ DEVICE_AND_API_INIT(gpio_nrf5_P0, CONFIG_GPIO_NRF5_P0_DEV_NAME, gpio_nrf5_P0_ini
 
-#endif /* CONFIG_GPIO_NRF5_P0 */
+static const struct gpio_nrf5_config gpio_nrf5_P1_cfg = {
+	.gpio_base_addr   = NRF_P1_BASE,
```
Unneeded space after .gpio_base_addr causing alignment issue with .gpio_port below.

mike-scott replied:
Unneeded space after .gpio_base_addr causing alignment issue with .gpio_port below.

mike-scott replied:
Unneeded space after .gpio_base_addr causing alignment issue with .gpio_port below.

mike-scott commented:
No idea how I got 3 comments that were the same at the end of this review :/

cvinayak committed:
drivers: gpio: nRF5: Use lowest IRQ priority level as default

Fix the permitted range of IRQ priority to, 0 to 2, as
maximum one priority level is used, for zero latency IRQs
in nRF51, being an ARM Cortex-M0.

Update the Kconfig default GPIOTE IRQ priority level to the
lowest, for nRF51 a level of 2; for nRF52, a level of 5.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

mike-scott reviewed (commented):
Code LGTM, I'll re-test locally in a bit (need to hook up my analyzer)

anangl commented on drivers/gpio/gpio_nrf5.c:
```diff
@@ -65,24 +65,22 @@ struct _gpiote {
 /** Configuration data */
 struct gpio_nrf5_config {
+	/* GPIO port */
+	u32_t gpio_port;
```
Is it worth spending `u32_t` to hold values 0-1?
I'd make it `u8_t` and move it to the end of the structure to avoid padding (which will probably appear anyway between structures for two ports when the linker places them one after another, but that's another story).

anangl commented:
LGTM. Just some minor comments.

cvinayak commented:
@anangl addressed your two comments.
